File: Examples/BioMechanics/Cxx/ParticleSpringSystem.cxx
#include "vtkRenderer.h"
#include "vtkRenderWindow.h"
#include "vtkRenderWindowInteractor.h"
#include "vtkPolyData.h"
#include "vtkPolyDataMapper.h"
#include "vtkPolyDataWriter.h"
#include "vtkDataSetMapper.h"
#include "vtkPoints.h"
#include "vtkCellArray.h"
#include "vtkIdList.h"
#include "vtkActor.h"
#include "vtkActor.h"
#include "vtkProperty.h"
#include "vtkCamera.h"

#include "vtkPolyData.h"
#include "vtkPolyDataWriter.h"
#include "vtkXMLPolyDataReader.h"

#include "vtkTimerLog.h"
#include "vtkDoubleArray.h"
#include "vtkPointLocator.h"

#include "vtkSmartPointer.h"
#include "vtkCommand.h"

#include "vtkCollision.h"
#include "vtkCollisionCollection.h"

#include "vtkPSSInterface.h"

class vtkPSSTimerCallback : public vtkCommand
{
public:
	static vtkPSSTimerCallback *New()
	{
		vtkPSSTimerCallback *cb = new vtkPSSTimerCallback;
		cb->FastTimerId = 0;
		cb->FasterTimerId = 0;
		cb->RenderTimerId = 0;
		return cb;
	}

	virtual void Execute(vtkObject *caller, unsigned long eventId, void *callData)
	{
		if (vtkCommand::TimerEvent == eventId)
		{
			int tid = * static_cast<int *>(callData);

			if (tid == this->FastTimerId)
			{
				cout << "new collision\n";

				vtkPolyData * mesh = vtkPolyData::SafeDownCast(this->DeformationModel->GetInput());

				//Locate collision points
				vtkPointLocator * locator = vtkPointLocator::New();
				double bounds[6];
				mesh->GetBounds(bounds);

				double p[3] = {bounds[0], 0, 0};

				locator->SetDataSet(mesh);

				vtkIdList * list = vtkIdList::New();
				vtkDoubleArray * directions = vtkDoubleArray::New();
				directions->SetNumberOfComponents(3);

				locator->FindClosestNPoints(5, p, list);

				//Set Collisions
				double dir[3];
				dir[0] = 0.2;//-0.1;
				dir[1] = 0.05;
				dir[2] = 0;//0.05;

				for(vtkIdType i = 0; i< list->GetNumberOfIds(); i++)
				{
					double * point = mesh->GetPoint(list->GetId(i));
					directions->InsertNextTuple(dir);
					vtkIdType id = list->GetId(i);

					//Insert collision info
					vtkCollision * collision = vtkCollision::New();
					collision->SetCollisionType(vtkCollision::ToolOrgan);
					collision->SetElementId(0, 0);
					collision->SetElementId(1, 0);

					//Organ cell point
					collision->SetPointId(1, id);
					collision->SetPoint(1, point);
					//collision->InsertCellId(0, organCellId);
					collision->SetDisplacement(dir);

					collision->Print(cout);

					this->DeformationModel->AddCollision(collision);
				}
			}
			else if (tid == this->FasterTimerId)
			{
				vtkTimerLog * timer = vtkTimerLog::New();
				timer->StartTimer();
				this->DeformationModel->Modified();
				this->DeformationModel->Update();
				timer->StopTimer();

				std::cout << "[Test] Execution Rate: " << 1/(timer->GetElapsedTime()) << "\n";

			}
			else if (tid == this->RenderTimerId)
			{
				vtkRenderWindowInteractor *iren = vtkRenderWindowInteractor::SafeDownCast(caller);
				if (iren && iren->GetRenderWindow() && iren->GetRenderWindow()->GetRenderers())
				{
					iren->Render();
				}
			}
		}
	}

	void SetFastTimerId(int tid)
	{
		this->FastTimerId = tid;
	}

	void SetFasterTimerId(int tid)
	{
		this->FasterTimerId = tid;
	}

	void SetRenderTimerId(int tid)
	{
		this->RenderTimerId = tid;
	}

	void SetDeformationModel(vtkPSSInterface * DeformationModel)
	{
		this->DeformationModel = DeformationModel;
	}

	void SetCollisionIds(vtkIdList * list)
	{
		this->List = list;
	}
private:
	int FastTimerId;
	int RenderTimerId;
	int FasterTimerId;

	vtkIdList * List;

	vtkPSSInterface * DeformationModel;
};

int main(int argc, char * argv[])
{
	const char * filename = "/home/jballesteros/Workspace/data/vtkESQuiData/Scenario/Organs/ball_def_c10.vtp";

	if (argc > 1)
	{
		filename = argv[1];
	}

	vtkSmartPointer<vtkXMLPolyDataReader> reader =
			vtkSmartPointer<vtkXMLPolyDataReader>::New();
	reader->SetFileName(filename);
	reader->Update();

	vtkPolyData * mesh = reader->GetOutput();

	vtkSmartPointer<vtkPSSInterface> pss = vtkSmartPointer<vtkPSSInterface>::New();
	pss->SetName("ball_def_c10");
	pss->SetFileName(filename);
	pss->SetColor(1,0,0);
	pss->SetOpacity(1.0);
	pss->SetSolverType(vtkMotionEquationSolver::VelocityVerlet);
	pss->SetSpringCoefficient(150);
	pss->SetDistanceCoefficient(10);
	pss->SetDampingCoefficient(5);//Friction
	pss->SetMass(.5);
	pss->SetDeltaT(0.001);//10ms
	pss->Init();

	vtkSmartPointer<vtkRenderer> renderer = vtkSmartPointer<vtkRenderer>::New();

	vtkSmartPointer<vtkRenderWindow> renWin =
			vtkSmartPointer<vtkRenderWindow>::New();
	renWin->SetSize(500,500);
	renWin->AddRenderer(renderer);

	vtkSmartPointer<vtkRenderWindowInteractor> iren =
			vtkSmartPointer<vtkRenderWindowInteractor>::New();
	iren->SetRenderWindow(renWin);

	vtkSmartPointer<vtkPolyDataMapper> mapper =
			vtkSmartPointer<vtkPolyDataMapper>::New();
	mapper->SetInput(mesh);

	vtkSmartPointer<vtkActor> actor =
			vtkSmartPointer<vtkActor>::New();
	actor->SetMapper(mapper);
	actor->GetProperty()->SetColor(0,1,0);
	actor->GetProperty()->SetOpacity(0.5);

	renderer->AddActor(actor);
	renderer->AddActor(pss->GetActor());
	renderer->SetBackground(1,1,1);

	renderer->ResetCamera();
	iren->Initialize();

	renWin->Render();

	// Sign up to receive TimerEvent:
	//
	vtkPSSTimerCallback * cb = vtkPSSTimerCallback::New();
	iren->AddObserver(vtkCommand::TimerEvent, cb);
	int tid;

	cb->SetDeformationModel(pss);

	//Create a faster timer for DeformationModel update
	tid = iren->CreateRepeatingTimer(50);
	cb->SetFasterTimerId(tid);

	//Create a collision every 5 seconds
	tid = iren->CreateRepeatingTimer(5000);
	cb->SetFastTimerId(tid);

	// Create a slower repeating timer to trigger Render calls.
	// (This fires at the rate of approximately 25 frames per second.)
	//
	tid = iren->CreateRepeatingTimer(25);
	cb->SetRenderTimerId(tid);

	iren->Start();

	return 0;
}

File: Examples/BioMechanics/Python/ParticleSpringSystem.py

import vtk
import vtkesqui
import time

class vtkTimerCallback():
   def __init__(self):
       self.timer_count = 0
 
   def execute(self,obj,event):
       #print self.timer_count
       
       
       start = time.time()
       #self.actor.SetPosition(self.timer_count, self.timer_count,0);
       self.ball.Modified()
       self.ball.Update()
       rate = 1/(time.time() - start)
       #print rate

       iren = obj
       iren.GetRenderWindow().Render()
       
       if (self.timer_count % 500 == 0):
         print "displacement"
         # Add a displacement
         self.defo.InsertDisplacement(18, 0.12, 0.02, 0.0)
         self.defo.InsertDisplacement(19, 0.25, 0.05, 0.0)
         self.defo.InsertDisplacement(20, 0.12, 0.02, 0.0)
         
       self.timer_count += 1

def main():

  fn ="/home/david/vtkESQuiData/Scenario/Organs/ball.vtp"
  tfn ="/home/david/vtkESQuiData/Scenario/Textures/muscle.jpg"
  cfn ="/home/david/vtkESQuiData/Scenario/Organs/ball_col.vtp"
  dfn ="/home/david/vtkESQuiData/Scenario/Organs/ball_def_c10.vtp"

  # Visualization Model
  vis = vtkesqui.vtkVisualizationModel()
  vis.SetName("ball_vis")
  vis.SetFileName(fn)
  vis.SetTextureFileName(tfn)
  vis.SetOpacity(1.0)
  vis.SetColor(1.0, 1.0, 1.0)

  # Visualization Model
  col = vtkesqui.vtkCollisionModel()
  col.SetName("ball_vtkbioeng")
  col.SetFileName(cfn)
  col.SetVisibility(1)
  
  # Deformation Model
  defo = vtkesqui.vtkParticleSpringSystemInterface()
  defo.SetName("ball_defo")
  defo.SetFileName(dfn)
  defo.SetOpacity(0.5)
  defo.SetVisibility(1)
  defo.SetColor(1.0, 1.0, 1.0)
  defo.SetSpring(150)
  defo.SetDistance(1.0)
  defo.SetDamping(3)
  defo.SetMass(.5)
  defo.SetTimeStep(0.001)

  ball = vtkesqui.vtkScenarioElement()
  ball.SetId(0)
  ball.SetName("ball")
  ball.SetPosition(3.0, 0.0, 0.0)
  ball.SetOrientation(0, 0, -20)
  ball.SetVisualizationModel(vis)
  ball.SetCollisionModel(col)
  ball.SetDeformationModel(defo)

  ball.Update()

  ren = vtk.vtkRenderer()
  renWin = vtk.vtkRenderWindow()
  renWin.AddRenderer(ren)
  renWin.SetWindowName("vtkESQui - Particle Spring System")
  iren = vtk.vtkRenderWindowInteractor()
  iren.SetRenderWindow(renWin)

  ren.AddActor(vis.GetActor())
  ren.AddActor(col.GetActor())
  ren.AddActor(defo.GetActor())

  ren.SetBackground(0.3, 0.6, 0.3)
  renWin.SetSize(800, 600)

  ren.ResetCamera()
  camera = ren.GetActiveCamera()
  camera.Azimuth(15)
  camera.Zoom(1.6)
  
  iren.Initialize()

  # Sign up to receive TimerEvent
  cb = vtkTimerCallback()
  cb.ball = ball
  iren.AddObserver('TimerEvent', cb.execute)
  timerId = iren.CreateRepeatingTimer(1)
  cb.renderTimerId = timerId
  cb.defo = defo
  
  iren.Start()
  
if __name__ == '__main__':
  main()
  
